R/moma_sfcca.R

```diff
@@ -0,0 +1,903 @@
+#' @param x_smooth,y_smooth An object of class inheriting from "\code{moma_smoothness_type}". Most conveniently
+#'          specified by functions described in \code{moma_smoothness}. It specifies the type of smoothness
+#'           terms used in the model. Note that for \code{moma_fcca}, these two parameters must not be
+#'          specified at the same time. For \code{moma_scca} and \code{moma_twscca}, they must not be specified.
+#' @param pg_settings An object of class inheriting from "\code{moma_pg_settings}". Most conviently
+#'          specified by functions described in \code{\link{moma_pg_settings}}. It specifies the type of algorithm
+#'          used to solve the problem, acceptable level of precision, and the maximum number of iterations allowed.
+#' @param max_bic_iter A positive integer. Defaults to 5. The maximum number of iterations allowed
+#' in nested greedy BIC selection scheme.
+#' @param rank A positive integer. Defaults to 1. The maximal rank, i.e., maximal number of principal components to be used.
+#' @export
+
+moma_sfcca <- function(X, ..., Y,
+                       center = TRUE, scale = FALSE,
+                       x_sparse = moma_empty(), y_sparse = moma_empty(),
+                       x_smooth = moma_smoothness(), y_smooth = moma_smoothness(),
+                       pg_settings = moma_pg_settings(),
+                       max_bic_iter = 5,
+                       rank = 1) {
+    chkDots(...)
+    error_if_not_of_class(x_sparse, "moma_sparsity_type")
+    error_if_not_of_class(y_sparse, "moma_sparsity_type")
+    error_if_not_of_class(x_smooth, "moma_smoothness_type")
+    error_if_not_of_class(y_smooth, "moma_smoothness_type")
+
+    return(SFCCA$new(
+        X,
+        Y = Y,
+        center = center, scale = scale,
+        # sparsity
+        x_sparsity = x_sparse$sparsity_type,
+        y_sparsity = y_sparse$sparsity_type,
+        lambda_x = x_sparse$lambda,
+        lambda_y = y_sparse$lambda,
+        # smoothness
+        Omega_x = x_smooth$Omega,
+        Omega_y = y_smooth$Omega,
+        alpha_x = x_smooth$alpha,
+        alpha_y = y_smooth$alpha,
+        pg_settings = pg_settings,
+        select_scheme_str = paste0( # the order is important
+            x_smooth$select_scheme,
+            y_smooth$select_scheme,
+            x_sparse$select_scheme,
+            y_sparse$select_scheme
+        ),
+        max_bic_iter = max_bic_iter,
+        rank = rank
+    ))
+}
+
+#' Perform one-way sparse CCA
+#'
+#' \code{moma_scca} is a function for performing one-way sparse CCA.
+#' @export
+#' @describeIn moma_sfcca a function for performing one-way sparse CCA.
+moma_scca <- function(X, ..., Y,
+                      center = TRUE, scale = FALSE,
+                      x_sparse = moma_empty(), y_sparse = moma_empty(),
+                      #    x_smooth = moma_smoothness(), y_smooth = moma_smoothness(),
```

Comment:
Every time I read this, it's still confusing to me: why is no smoothing the default if we recommend SFPCA as a general case?

Reply:
`moma_scca` is supposed to impose only sparsity, isn't it?

What happens in `moma_smoothness` is if user specifies alpha explicitly but not Omega, then we use second different matrix as smoothing.

Also please check out this test case, 

https://github.com/Banana1530/MoMA-1/blob/64a34bbad62aa3dc2c553a70dc8e9be980477041/tests/testthat/test_sfpca_wrapper.R#L647

and the latest commit "Give info: use sec-diff-mat as default":
```R
check_omega <- function(Omega, alpha, n) {

    ## LOGIC:
    # if alpha = 0: overwrite Omega_u to identity matrix whatever it was
    # if alpha is a grid or a non-zero scalar: 
    #       if Omega missing: set to second-difference matrix
    #       else check validity
...
}
```
In conclusion, if the user specifies alpha, he gets sec-diff-mat or whatever he wants as the smoothing matrix; if he does NOT specify alpha, then he gets no smoothing.

Reply:
second different matrix -> second difference matrix